recurring_contract/models/move.py

```diff
@@ -178,8 +177,101 @@ def find_sum(numbers, target, partial=None):
+        """
+        # Ensure that the function receive only one invoice
+        self.ensure_one()
+        # Build the dictionnary
+        inv_val_dict = dict()
+        inv_val_dict[self.name] = {
+            'invoice_line_ids': self._build_invoice_lines(inv_lines) if inv_lines else self.invoice_line_ids.ids,
+            'date': due_date if due_date else self.date,
+            'payment_reference': ref if ref else self.payment_reference,
+            'payment_mode_id': pay_mode_id if pay_mode_id else self.payment_mode_id.id
+        }
+        return inv_val_dict
+
+    def _build_invoice_lines(self, inv_lines):
+        """
+        It creates a list of tuples that will be used to create, modify or delete invoice lines.
+
+        :param inv_lines: a dictionary of the form {product_id: {'amt': amount, 'contract': contract_id}}
+        :return: A list of tuples.
+        """
+        res = list()
+        # create or modify existing line
+        for product, dict_data in inv_lines.items():
+            amt = dict_data.get("amt")
+            if amt == 0:
+                res.append("to_cancel")
+            else:
+                line_ids = self.invoice_line_ids.filtered(lambda l: l.product_id == product).ids
+                if not line_ids:
+                    contract = dict_data.get("contract")
+                    contract_line = contract.contract_line_ids.filtered(lambda l: l.product_id == product)
+                    res.append((0, 0,
+                                {
+                                    "price_unit": amt,
+                                    "product_id": product.id,
+                                    'name': product.name,
+                                    'quantity': contract_line.quantity,
+                                    'contract_id': contract_line.contract_id.id,
+                                    'account_id': product.with_company(
+                                        contract_line.contract_id.company_id.id).property_account_income_id.id or False
+                                }
+                                ))
+                else:
+                    for line_id in line_ids:
+                        res.append((1, line_id, {"price_unit": amt}))
+
+        # Delete line if the product shouldn't be invoiced anymore
+        if len(inv_lines) < len(self.invoice_line_ids):
+            lines_to_delete = self.invoice_line_ids.ids
+            for line_id in self.invoice_line_ids.ids:
+                for tuple in res:
+                    if tuple[1] == line_id:
+                        lines_to_delete.remove(line_id)
+            for line_id_to_del in lines_to_delete:
+                res.append((2, line_id_to_del, 0))
+        return res
+
+    def _cancel_invoice(self):
+        """
+        It cancels the invoice
+        """
+        self.ensure_one()
+        self.button_cancel()
```

Comment:
I don't think this method brings any value. This could be done outside of this function, no?